### PR TITLE
metadata(cdc): add sql implementation (part 1)

### LINKED ISSUE #9767: Implement sql based metadata

### PR DESCRIPTION
**Pull request by CharlesCheung96**
### What problem does this PR solve?

Issue Number: close #9767

### What is changed and how it works?
1. Add metadata client interface to decouple the upper interface from different storage implementations(such as; sql, orm and in-memory storage).
2. Implement sql-based client interface.

#### Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)
 - No code

### Release note 

```release-note
Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

If you don't think this PR needs a release note then fill it with `None`.
```

### CODE REVIEW

**Review by sdojjy — APPROVED** [review]

File: cdcv2/metadata/sql/client.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+	"gorm.io/gorm"
+)
+
+// TxnAction is a series of operations that can be executed in a transaction and the
+// generic type T represents the transaction context.
+//
+// Note that in the current implementation the metadata operation and leader check are
+// always in the same transaction context. In the future, cross-database transactions
+// could also be supported with different implementations.
```
Comment (asddongmen) [inline cdcv2/metadata/sql/client.go:29]: What does `cross-database transactions` mean? Why we need this?

File: cdcv2/metadata/sql/model.go [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+func (cp *CaptureProgress) Scan(value interface{}) error {
+	b, ok := value.([]byte)
+	if !ok {
+		return errors.New("type assertion to []byte failed")
+	}
+
+	return json.Unmarshal(b, cp)
+}
+
+// ReplicaConfig stores the configuration for a replica.
```
Comment (asddongmen) [inline cdcv2/metadata/sql/model.go:65]: please correct this comment

**Review by asddongmen — APPROVED** [review]

File: cdcv2/metadata/sql/client.go (hunk 1 above)
Comment (CharlesCheung96) [reply cdcv2/metadata/sql/client.go:29]: The comments here are somewhat ambiguous. In fact, there is no need for cross-database transactions in the current implementation. And i will remove such comments.

File: cdcv2/metadata/sql/model.go (hunk 2 above)
Comment (CharlesCheung96) [reply cdcv2/metadata/sql/model.go:65]: ReplicaConfig will be removed in the next PR.

**Pull request merged**